pennylane/templates/embeddings/qaoa.py

```diff
@@ -260,22 +259,38 @@ def circuit(weights, f=None):
         1-dimensional Ising model.
 
     """
-    wires = Wires(wires)
-    repeat = _preprocess(features, wires, weights)
-
-    if local_field == "Z":
-        local_fields = RZ
-    elif local_field == "X":
-        local_fields = RX
-    elif local_field == "Y":
-        local_fields = RY
-    else:
-        raise ValueError(f"did not recognize local field {local_field}")
 
-    for l in range(repeat):
-        # apply alternating Hamiltonians
-        qaoa_feature_encoding_hamiltonian(features, wires)
-        qaoa_ising_hamiltonian(weights[l], wires, local_fields)
+    num_params = 2
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, features, weights, wires, local_field="Y", do_queue=True):
+
+        wires = Wires(wires)
+
+        _preprocess(features, wires, weights)
+
+        if local_field == "Z":
+            self.local_fields = RZ
+        elif local_field == "X":
+            self.local_fields = RX
+        elif local_field == "Y":
+            self.local_fields = RY
+        else:
+            raise ValueError(f"did not recognize local field {local_field}")
 
-    # repeat the feature encoding once more at the end
-    qaoa_feature_encoding_hamiltonian(features, wires)
+        super().__init__(features, weights, wires=wires, do_queue=do_queue)
+
+    def decomposition(self, features, weights, wires):
+
+        # first dimension of the weights tensor determines
+        # the number of layers
+        repeat = qml.math.shape(weights)[0]
+
+        for l in range(repeat):
+            # apply alternating Hamiltonians
+            qaoa_feature_encoding_hamiltonian(features, wires)
+            qaoa_ising_hamiltonian(weights[l], wires, self.local_fields)
+
+        # repeat the feature encoding once more at the end
+        qaoa_feature_encoding_hamiltonian(features, wires)
```

Comment:
Just a small comment here: once #1100 is merged and the new queuing system fully replaces the old queuing system, there will be two changes required:

- the method should be renamed `expand` 
- it should return a tape containing the operations in addition

This is a relatively minor change, basically it becomes:

```python
with QuantumTape() as tape:
        for l in range(repeat):
            # apply alternating Hamiltonians
            qaoa_feature_encoding_hamiltonian(features, wires)
            qaoa_ising_hamiltonian(weights[l], wires, self.local_fields)

        # repeat the feature encoding once more at the end
        qaoa_feature_encoding_hamiltonian(features, wires)

return tape
```

Reply:
Will implement this as of now. 

I see, currently the `expand` function just makes a tape from `decomposition`, but in future we will likely have `expand` being a method of each operation?

Reply:
Yep - currently, for unitary operations, `expand` simply uses `decomposition`, but post #1100 the idea is that all operations have a native `expand`, that makes sense for that operation.

E.g., an observable would expand to give the diagonalization gates